instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/TracingExecutionInterceptor.java

```diff
@@ -289,6 +303,43 @@ private void onSdkResponse(
     }
   }
 
+  @Override
+  public void afterTransmission(
+      Context.AfterTransmission context, ExecutionAttributes executionAttributes) {
+    if (recordIndividualHttpError) {
+      extractHttpErrorAsEvent(context, executionAttributes);
+    }
+  }
+
+  private static void extractHttpErrorAsEvent(
+      Context.AfterTransmission context, ExecutionAttributes executionAttributes) {
+    io.opentelemetry.context.Context otelContext = getContext(executionAttributes);
+    if (otelContext != null) {
+      Span span = Span.fromContext(otelContext);
+      SdkHttpResponse response = context.httpResponse();
+
+      if (!response.isSuccessful()) {
+        int errorCode = response.statusCode();
+        // we want to record the error message from http response
+        Optional<InputStream> responseBody = context.responseBody();
+        if (responseBody.isPresent()) {
+          String errorMsg =
+              new BufferedReader(
+                      new InputStreamReader(responseBody.get(), Charset.defaultCharset()))
```

Comment:
Is the underlying `InputStream` still available for use by the SDK after calling `responseBody.get()`?

Reply:
Thank you for catching this! 

I think there is no guarantee that the InputStream will support `mark()/reset()` operations or seeking. So the next time if the same InputStream is read, it will not give you anything. 

To verify, I did an simple experiment. Based on the pipeline stage in aws java sdk: https://github.com/aws/aws-sdk-java-v2/blob/d020d37138eee9d4d74e814086143d26d923fee0/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AfterTransmissionExecutionInterceptorsStage.java#L43-L46, it seems like `modifyHttpResponse` call is after `afterTransmission` call. So I tested the same InputStream in `modifyHttpResponse` hooks in ExecutionInterceptor and confirmed it's already empty (after it's read out in `afterTransmission`). 

It's a valid concern that the inputStream for the response body will become useless after we read it out in `afterTransmission` hook. Fortunately there is another hook `Optional<InputStream> modifyHttpResponseContent` which allow us to modify the content of the a response. We can copy the content out and generate a new InputStream from it. A similar approach has already been implemented in the aws java adk: https://github.com/aws/aws-sdk-java-v2/blob/d020d37138eee9d4d74e814086143d26d923fee0/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/interceptor/HttpChecksumValidationInterceptor.java#L55-L73

I will make the change in next commit.

Reply:
Thanks @pxaws

Is there a way to verify in the test that the `InputStream` is available to the client and has the expected value?

Reply:
Yes, I modified the test to do that. I basically defined another execution interceptor which is registered with AWS SDK after the TracingExecutionInterceptor used for SDK instrumentation and then verify from there to make sure the response body contains the expected content. Please see the next commit. Thank you!